programs/blockvaultd/log_store.cpp

```diff
@@ -0,0 +1,279 @@
+#include "log_store.hpp"
+#include <cassert>
+#include <iostream>
+
+namespace blockvault {
+
+using namespace nuraft;
+namespace fs = std::filesystem;
+
+ptr<log_entry> empty_log_entry() {
+   static ptr<buffer>    buf = buffer::alloc(sz_ulong);
+   static ptr<log_entry> r   = cs_new<log_entry>(0, buf);
+   return r;
+}
+
+fs::path path_for_index(const fs::path& log_dir, ulong index) {
+   // the log is stored in log_dir/<log_index>.log
+   return (log_dir / std::to_string(index)).concat(".log");
+}
+
+fs::path store_log(const fs::path& log_dir, const unsigned char* data, int sz) {
+   namespace fs = std::filesystem;
+
+   auto file = log_dir / std::tmpnam(nullptr);
+   auto fp   = std::unique_ptr<FILE, decltype(&fclose)>{fopen(file.c_str(), "w"), &fclose};
+   if (!fp)
+      throw fs::filesystem_error("file open error", file, std::make_error_code(static_cast<std::errc>(errno)));
+   fwrite(data, 1, sz, fp.get());
+   return file;
+}
+
+struct log_entry_meta {
+   ulong term;
+   byte  type;
+
+   log_entry_meta(ulong tm = 0, log_val_type tp = log_val_type::app_log)
+       : term(tm)
+       , type(static_cast<byte>(tp)) {}
+};
+
+constexpr int log_entry_meta_size = sizeof(ulong) + sizeof(byte);
+
+fs::path store_log(const fs::path& log_dir, nuraft::ptr<nuraft::log_entry>& entry) {
+   namespace fs = std::filesystem;
+
+   auto file = log_dir / std::tmpnam(nullptr);
+   auto fp   = std::unique_ptr<FILE, decltype(&fclose)>{fopen(file.c_str(), "w"), &fclose};
+   if (!fp)
+      throw fs::filesystem_error("file open error", file, std::make_error_code(static_cast<std::errc>(errno)));
+
+   log_entry_meta meta{entry->get_term(), entry->get_val_type()};
+
+   fwrite(&meta, log_entry_meta_size, 1, fp.get());
+   auto& buf = entry->get_buf();
+   buf.pos(0); // reset the position
+   fwrite(buf.data(), 1, buf.size(), fp.get());
+   return file;
+}
+
+class log_entry_reader {
+ public:
+   log_entry_reader() = default;
+   log_entry_reader(fs::path log_dir, ulong index) { set(log_dir, index); }
+   ~log_entry_reader() {
+      if (!fname.empty()) {
+         std::error_code ec;
+         fs::remove(fname, ec);
+      }
+   }
+
+   log_entry_reader(const log_entry_reader&) = delete;
+   log_entry_reader& operator=(const log_entry_reader&) = delete;
+   log_entry_reader(log_entry_reader&& other) { fname.swap(other.fname); }
+
+   void set(fs::path log_dir, ulong index) {
+      auto original = path_for_index(log_dir, index);
+      if (fs::exists(original)) {
+         fname = log_dir / std::tmpnam(nullptr);
+         fs::create_hard_link(path_for_index(log_dir, index), fname);
+      }
+   }
+
+   ptr<log_entry> get() {
+      auto fp = std::unique_ptr<FILE, decltype(&fclose)>(fopen(fname.c_str(), "r"), &fclose);
```

Comment:
can fc::cfile be used?

Reply:
At the point, I try to avoid anything requires me to change the NuRaft source code. The code is executed from the thread created by NuRaft which doesn't handle fc::exceptions which fc::file throws.